src/paths/widest_paths.c

```diff
@@ -0,0 +1,747 @@
+ *        \c to are reached.
+ * \param inbound_edges A pointer to an initialized igraph vector or null.
+ *        If not null, a vector containing the inbound edge of each vertex in
+ *        the single source widest path tree is returned here. The
+ *        inbound edge of vertex i in the tree is the edge via which vertex i
+ *        was reached. The start vertex and vertices that were not reached
+ *        during the search will have -1 in the corresponding entry of the
+ *        vector. Note that the search terminates if all the vertices in
+ *        \c to are reached.
+ * \return Error code:
+ *        \clist
+ *        \cli IGRAPH_ENOMEM
+ *           not enough memory for temporary data.
+ *        \cli IGRAPH_EINVVID
+ *           \p from is invalid vertex ID, or the length of \p to is
+ *           not the same as the length of \p vertices (if not NULL) or
+ *           the length of \p edges (if not NULL)
+ *        \cli IGRAPH_EINVMODE
+ *           invalid mode argument.
+ *        \endclist
+ *
+ * Time complexity: O(|E|log|E|+|V|), where |V| is the number of
+ * vertices and |E| is the number of edges
+ *
+ * \sa \ref igraph_widest_paths_dijkstra() or \ref igraph_widest_paths_floyd_warshall()
+ * if you only need the path length but not the paths themselves.
+ */
+igraph_error_t igraph_get_widest_paths(const igraph_t *graph,
+                                       igraph_vector_ptr_t *vertices,
+                                       igraph_vector_ptr_t *edges,
+                                       igraph_integer_t from,
+                                       igraph_vs_t to,
+                                       const igraph_vector_t *weights,
+                                       igraph_neimode_t mode,
+                                       igraph_vector_int_t *predecessors,
+                                       igraph_vector_int_t *inbound_edges) {
+
+    /* Implementation details: This is a Dijkstra algorithm with a
+    binary heap, modified to support widest paths. The heap is indexed,
+    so it stores both the widest path to a node, as well as it's index. We
+    use a 2 way heap so that we can query indexes directly in the heap.
+
+    To adapt a Dijkstra to handle widest path, instead of prioritising candidate
+    nodes with the minimum distance, we prioritise those with the maximum
+    width instead. When adding a node into our set of 'completed' nodes, we
+    update all neighbouring nodes with a width that is equal to the min of the
+    width to the current node and the width of the edge.
+
+    We denote the widest path from a node to itself as infinity, and the widest
+    path from a node to a node it cannot reach as negative infinity.
+    */
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph);
+    igraph_integer_t no_of_edges = igraph_ecount(graph);
+    igraph_real_t my_posinfinity = IGRAPH_POSINFINITY;
+    igraph_real_t my_neginfinity = IGRAPH_NEGINFINITY;
```

Comment:
why do you rename these?

Reply:
I guess this pattern was lifted from the shortest path functions. I'm also not sure why it was done this way there. My guess is that in the past, `IGRAPH_POSINFINITY` may have been defined through a function, i.e. evaluation was expensive:

https://github.com/igraph/igraph/blob/0.8.5/include/igraph_types.h#L68

@gaborcsardi Do you remember why it was done this way?

Reply:
A good chunk of the code is copied from the equivalent function in `src/paths/dijkstra.c`. Specifically:

- `igraph_widest_paths_dijkstra` is `igraph_shortest_paths_dijkstra` 
- `igraph_get_widest_paths` is `igraph_get_shortest_paths_dijkstra`
- `igraph_get_widest_path` is `igraph_get_shortest_path_dijkstra`

I was originally planning on writing my own faster functions to replace dijkstra, but realised that the difference in run time in dijkstra was already good enough for most use cases and there was no need for me to reinvent the wheel.

Reply:
It makes sense that it used to be expensive because it was a function call. Let's now use IGRAPH_POSINFINITY etc. directly. They can also be changed in `src/paths/dijkstra.c`.